cryptography/src/lib.rs

```diff
@@ -38,57 +36,8 @@ pub enum Error {
     InvalidSignature,
     #[error("invalid signature length")]
     InvalidSignatureLength,
-    #[error("invalid bytes")]
-    InsufficientBytes,
-}
-
-/// Types that can be fallibly read from a fixed-size byte sequence.
-///
-/// `Array` is typically used to parse things like `PublicKeys` and `Signatures`
-/// from an untrusted network connection. Once parsed, these types are assumed
-/// to be well-formed (which prevents duplicate validation).
-///
-/// If a byte sequencer is not properly formatted, `TryFrom` must return an error.
-pub trait Array:
-    Clone
-    + Send
-    + Sync
-    + 'static
-    + Eq
-    + PartialEq
-    + Ord
-    + PartialOrd
-    + Debug
-    + Hash
-    + Display
-    + AsRef<[u8]>
-    + Deref<Target = [u8]>
-    + for<'a> TryFrom<&'a [u8], Error = Error>
-    + for<'a> TryFrom<&'a Vec<u8>, Error = Error>
-    + TryFrom<Vec<u8>, Error = Error>
-    + SizedSerialize
-{
-    /// Attempts to read an array from the provided buffer.
-    fn read_from<B: Buf>(buf: &mut B) -> Result<Self, Error> {
-        // Check if there are enough bytes in the buffer to read a digest.
-        let len = Self::SERIALIZED_LEN;
-        if buf.remaining() < len {
-            return Err(Error::InsufficientBytes);
-        }
-
-        // If there are enough contiguous bytes in the buffer, use them directly.
-        let chunk = buf.chunk();
-        if chunk.len() >= len {
-            let array = Self::try_from(&chunk[..len])?;
-            buf.advance(len);
-            return Ok(array);
-        }
-
-        // Otherwise, copy the bytes into a temporary buffer.
-        let mut temp = vec![0u8; len];
-        buf.copy_to_slice(&mut temp);
-        Self::try_from(temp)
-    }
+    #[error("array error: {0}")]
+    ArrayError(#[from] array::Error),
```

Comment:
Still not entirely sure why this is needed since I don't actually seem to use the "ArrayError" defind here anywhere, but without this the "type Error = Error" in all the classes implementing Array resulted in errors.

Reply:
That is odd. Hm.

Reply:
It must be related to the fact that Array::read_from can return Error::InsufficientBytes.  Though it's read_from that calls try_from, not the other way around, so I'm still not quite fully understanding why there needs to be a conversion in this direction.

Reply:
Ah it's because I had to add From<Error> to the type constraint:

`type Error: std::error::Error + Send + Sync + 'static + From<Error>;`

So every implementor of Array needs to be able to convert its errors to array::Error.    If I try to get rid of the From<Error> constraint then read_from imlementation then I can't get it to return array::Error::InsufficientBytes

Reply:
What you figured out makes sense, I just think its annoying to require every implementer to have to support this for a single default function on the trait. 🤔 

Proposing we consider this: https://github.com/commonwarexyz/monorepo/pull/516